File: app/src/main/java/gc/david/dfm/faq/GetFaqsInteractor.java
/*
 * Copyright (c) 2017 David Aguiar Gonzalez
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package gc.david.dfm.faq;

import java.util.Set;

import gc.david.dfm.executor.Executor;
import gc.david.dfm.executor.Interactor;
import gc.david.dfm.executor.MainThread;
import gc.david.dfm.faq.model.Faq;

/**
 * Created by david on 17.12.16.
 */
public class GetFaqsInteractor implements Interactor, GetFaqsUseCase {

    private final Executor          executor;
    private final MainThread        mainThread;
    private final GetFaqsRepository repository;

    private Callback callback;

    public GetFaqsInteractor(final Executor executor, final MainThread mainThread, final GetFaqsRepository repository) {
        this.executor = executor;
        this.mainThread = mainThread;
        this.repository = repository;
    }

    @Override
    public void execute(final Callback callback) {
        if (callback == null) {
            throw new IllegalArgumentException("Callback can't be null, the client of this interactor needs to get the response in the callback");
        }
        this.callback = callback;
        this.executor.run(this);
    }

    @Override
    public void run() {
        try {
            final Set<Faq> faqs = repository.getFaqs();

            mainThread.post(new Runnable() {
                @Override
                public void run() {
                    callback.onFaqsLoaded(faqs);
                }
            });
        } catch (Exception exception) {
            mainThread.post(new Runnable() {
                @Override
                public void run() {
                    callback.onError();
                }
            });
        }
    }
}

File: app/src/main/java/gc/david/dfm/faq/GetFaqsInteractor.kt
/*
 * Copyright (c) 2021 David Aguiar Gonzalez
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package gc.david.dfm.faq

import gc.david.dfm.executor.Interactor
import gc.david.dfm.executor.NewMainThread
import gc.david.dfm.executor.NewThreadExecutor
import gc.david.dfm.faq.model.Faq

/**
 * Created by david on 17.12.16.
 */
class GetFaqsInteractor(
        private val executor: NewThreadExecutor,
        private val mainThread: NewMainThread,
        private val repository: NewGetFaqsDiskDataSource
) : Interactor {

    private lateinit var callback: Callback

    fun execute(callback: Callback) {
        this.callback = callback
        this.executor.run(this)
    }

    override fun run() {
        try {
            val faqs = repository.getFaqs()

            mainThread.post(Runnable { callback.onFaqsLoaded(faqs) })
        } catch (exception: Exception) {
            mainThread.post(Runnable { callback.onError() })
        }
    }

    interface Callback {

        fun onFaqsLoaded(faqs: Set<Faq>)

        fun onError()

    }
}
